File: julia/GradBench/src/benchmarks/llsq.jl
module LLSQ

import GradBench

struct Input
    x::Vector{Float64}
    n::Int64
end

abstract type AbstractLLSQ <: GradBench.Experiment end

function GradBench.preprocess(::AbstractLLSQ, message)
    x = convert(Vector{Float64}, message["x"])
    n = message["n"]
    return (Input(x, n),)
end

function t(i::Int64, n::Int64)
    return -1 + (Float64(i) * 2) / (Float64(n) - 1)
end

function primal(x::Vector{T}, n::Int64) where {T}
    m = length(x)

    f(i) = begin
        ti = t(i, n)
        inner_sum = 0.0
        mul = 1.0
        for j in 1:m
            inner_sum += -x[j] * mul
            mul *= ti
        end
        return (sign(ti) + inner_sum)^2
    end

    return sum(f(i) for i in 0:n-1) / 2
end

struct PrimalLLSQ <: AbstractLLSQ end
function (::PrimalLLSQ)(input)
    return primal(input.x, input.n)
end


end

File: python/gradbench/gradbench/tools/jax/llsq.py
import sys
import jax
import jax.numpy as jnp
import numpy as np
from gradbench import wrap
from functools import partial

jax.config.update("jax_enable_x64", True)


def t(i, n):
    return -1.0 + (i * 2.0) / (n - 1.0)


@partial(jax.jit, static_argnums=[1])
def llsq(x, n):
    m = x.shape[0]

    def f(i):
        ti = t(i, n)
        muls = jnp.full(m, ti).at[0].set(1)
        muls = jnp.cumprod(muls)
        g = lambda mul, xj: -(xj * mul)
        g_vals = jnp.vectorize(g)(muls, x)
        return (jnp.sign(ti) + jnp.sum(g_vals)) ** 2

    i_vals = jnp.arange(n)
    results = jax.vmap(f)(i_vals)

    return jnp.sum(results) / 2.0


@partial(jax.jit, static_argnums=[1])
def grad_llsq(x, n):
    return jax.grad(llsq, argnums=0)(x, n)


def prepare_input(input):
    return (np.array(input["x"], dtype=np.float64), input["n"])


@wrap.multiple_runs(pre=prepare_input, post=lambda x: float(x))
def primal(input):
    x, n = input
    return llsq(x, n)


@wrap.multiple_runs(pre=prepare_input, post=lambda x: x.tolist())
def gradient(input):
    x, n = input
    return grad_llsq(x, n)
